internal/orchestrator/updates.go

```diff
@@ -47,58 +47,109 @@ func (o *Orchestrator) startUpdateListener(ctx context.Context) {
 		select {
 		case <-ctx.Done():
-			o.logger.Info("stopping KV update listener")
-			return
-			// retrieve and process events sent by controllers.
-		case entry := <-installWatcher.Updates():
-			if entry == nil {
-				o.logger.Debug("nil kv entry")
-				continue
-			}
-			evt, err = installEventFromKV(ctx, entry)
-			if err != nil {
-				o.logger.WithError(err).Warn("error creating install condition event")
-				continue
+			o.logger.Debug("stopping KV update listener")
+			stop = true
+
+		// retrieve and process events sent by controllers.
+		case evt := <-evtChan:
+			le := o.logger.WithFields(logrus.Fields{
+				"conditionID":    evt.ConditionUpdate.ConditionID.String(),
+				"conditionState": string(evt.ConditionUpdate.State),
+				"kind":           string(evt.Kind),
+			})
+
+			if err := o.eventHandler.UpdateCondition(ctx, evt); err != nil {
+				le.WithError(err).Warn("updating condition")
 			}
 
-			/* XXX: Uncomment this for out-of-band inventory support
-			case entry := <-inventoryWatcher.Updates():
-				evt, err := inventoryEventFromKV(entry)
-				if err != nil {
-					o.logger.WithError(err).Warn("error creating inventory condition event")
-				} */
+			if err := o.notifier.Send(evt); err != nil {
+				le.WithError(err).Warn("sending notification")
+			}
 		}
-		le := o.logger.WithFields(logrus.Fields{
-			"conditionID":    evt.ConditionUpdate.ConditionID.String(),
-			"conditionState": string(evt.ConditionUpdate.State),
-			"kind":           string(evt.Kind),
-		})
+	}
+
+	wg.Wait()
+	close(evtChan)
 
-		if err := o.eventHandler.UpdateCondition(ctx, evt); err != nil {
-			le.WithError(err).Warn("updating condition")
+	o.logger.Debug("shut down KV updates")
+}
+
+type translatorFn func(context.Context, nats.KeyValueEntry) (*v1types.ConditionUpdateEvent, error)
+
+// startConditionListeners does what it says on the tin; iterate across all configured conditions
+// and start a KV watcher for each one. We increment the waitgroup counter for each condition we
+// can handle.
+func (o *Orchestrator) startConditionListeners(ctx context.Context,
+	evtChan chan<- *v1types.ConditionUpdateEvent, wg *sync.WaitGroup) {
+
+	for _, def := range o.conditionDefs {
+		var tf translatorFn
+		var watcher nats.KeyWatcher
+
+		kind := def.Kind
+
+		switch kind {
+		case ptypes.FirmwareInstall:
+			tf = installEventFromKV
+		case ptypes.InventoryOutofband:
+			tf = inventoryEventFromKV
+		default:
+			o.logger.WithField("condition.kind", string(kind)).Warn("unsupported condition")
+			continue
 		}
 
-		if err := o.notifier.Send(evt); err != nil {
-			le.WithError(err).Warn("sending notification")
+		wg.Add(1)
+
+		watcher, err := status.WatchConditionStatus(ctx, kind, o.facility)
+		if err != nil {
+			o.logger.WithError(err).WithField("condition.kind", string(kind)).Fatal("unable to get watcher")
 		}
+
+		go func() {
+			defer wg.Done()
+			// remember, the condition has to be true to run the loop, so "not stop" == false
```

Comment:
we can't use a break because it's in a `select` and that will take precedence (and do nothing). We could use a label and `goto` and I think that's as-ugly (plus potentially adds a headache for maintenance and extension).